libsolidity/analysis/ImmutableValidator.cpp

```diff
@@ -0,0 +1,271 @@
+		if ((
+			funcType->kind() == FunctionType::Kind::Internal ||
+			funcType->kind() == FunctionType::Kind::Declaration
+		) && funcType->hasDeclaration())
+			visitCallable(funcType->declaration());
+
+	return false;
+}
+
+bool ImmutableValidator::visit(IfStatement const& _ifStatement)
+{
+	bool prevInBranch = m_inBranch;
+
+	_ifStatement.condition().accept(*this);
+
+	m_inBranch = true;
+	_ifStatement.trueStatement().accept(*this);
+
+	if (auto falseStatement = _ifStatement.falseStatement())
+		falseStatement->accept(*this);
+
+	m_inBranch = prevInBranch;
+
+	return false;
+}
+
+bool ImmutableValidator::visit(WhileStatement const& _whileStatement)
+{
+	bool prevInLoop = m_inLoop;
+	m_inLoop = true;
+
+	_whileStatement.condition().accept(*this);
+	_whileStatement.body().accept(*this);
+
+	m_inLoop = prevInLoop;
+
+	return false;
+}
+
+bool ImmutableValidator::visit(Identifier const& _identifier)
+{
+	if (auto const callableDef = dynamic_cast<CallableDeclaration const*>(_identifier.annotation().referencedDeclaration))
+		visitCallable(*findFinalOverride(callableDef));
+	if (auto const varDecl = dynamic_cast<VariableDeclaration const*>(_identifier.annotation().referencedDeclaration))
+		analyseVariableDeclaration(*varDecl, _identifier);
+
+	return false;
+}
+
+bool ImmutableValidator::visit(Return const& _return)
+{
+	if (m_currentConstructor == nullptr)
+		return true;
+
+	if (auto retExpr = _return.expression())
+		retExpr->accept(*this);
+
+	checkAllVariablesInitialized(_return.location());
+
+	return false;
+}
+
+bool ImmutableValidator::analyseCallable(CallableDeclaration const& _callableDeclaration)
+{
+	FunctionDefinition const* prevConstructor = m_currentConstructor;
+	m_currentConstructor = nullptr;
+
+	if (FunctionDefinition const* funcDef = dynamic_cast<decltype(funcDef)>(&_callableDeclaration))
+	{
+		if (funcDef->isConstructor())
+			m_currentConstructor = funcDef;
+
+		// Disallow init. in the args of mods/base'ctor calls
+		bool previousInitAllowed = m_initializationOfImmutableAllowed;
+		m_initializationOfImmutableAllowed = false;
+
+		ASTNode::listAccept(funcDef->modifiers(), *this);
+
+		m_initializationOfImmutableAllowed = previousInitAllowed;
+
+		if (funcDef->isImplemented())
+			funcDef->body().accept(*this);
+	}
+	else if (ModifierDefinition const* modDef = dynamic_cast<decltype(modDef)>(&_callableDeclaration))
+		modDef->body().accept(*this);
+
+	m_currentConstructor = prevConstructor;
+
+	return false;
+}
+
+void ImmutableValidator::analyseVariableDeclaration(VariableDeclaration const& _variableDeclaration, Expression const& _expression)
+{
+	if (!_variableDeclaration.isStateVariable() || !_variableDeclaration.immutable())
+		return;
+
+	if (_expression.annotation().lValueRequested && _expression.annotation().lValueOfOrdinaryAssignment)
+	{
+		if (!m_currentConstructor)
+			m_errorReporter.typeError(
+				_expression.location(),
+				"Immutable variables can only be initialized inline or directly in the constructor."
```

Comment:
```suggestion
				"Immutable variables can only be initialized inline or assigned directly in the constructor."
```

Reply:
In case this is a regular assignment, people might not see the connection to the place of assignment. ("Yes, I know they can only be initialized in the constructor, but I'm not initializing, I'm assigning)